### PR TITLE
Expose theme `font_size` as an `Appearance` setting

### LINKED ISSUE #6109: Expose the font-size theme value in the Preferences
This issue tracks work to expose the font-size theme value in the Preferences.

Context/previous work:
> @dalthviz I think everything looks in order now for this?
> Then, once merged:
> 
> *  we need an issue to expose the font-size theme value in the Preferences?
> *  we change the npe2 PR from draft?

I think is ready @psobolewskiPhD ! And the next steps sound good to me :+1:

_Originally posted by @dalthviz in https://github.com/napari/napari/issues/5607#issuecomment-1634432213_

### PR DESCRIPTION
**Pull request by dalthviz**
# Fixes/Closes

Closes #6109

# Description

* Shows a widget to control the font size setting in the Appearance config page in the preferences dialog
* Connect settings changes to widget changes (to preserver theme default font size vs settings font size when changing themes):
    * If current theme font size corresponds to the font size from the settings the new theme default font size is used
* Set a minimum size for the preferences dialog.
* Add scrollbars to each preference page so the font size changes don't change the size of the preferences dialog

A preview:

theme_toggle

## Notes:

* ~The changes to prevent weird resizing while changing the `font_size` config causes the dialog to stay at a fixed size based on the child widgets size hint. For more info see https://doc.qt.io/qt-6/qlayout.html#SizeConstraint-enum~ Changed the approach to prevent the weird resizing by adding scrollbars to the preferences pages and setting a minimum size for the dialog.
* Not sure if a custom widget is needed for the new setting or if there is another approach to follow when adding new settings.
* ~Not sure if this requires a manual update of the documentation related with preferences (I think a page exists that showcases the preferences available but not totally sure if that is autogenerated 🤔 )~ Checking the docs preview, a new entry for the setting is there :+1:

## Type of change

- [x] Enhancement (non-breaking improvements of an existing feature)

# How has this been tested?

- [x] all tests pass with my change
- [x] I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality

### CODE REVIEW

**Comment by brisvag** [comment]
Those auto-resizings are a bit odd O.o

**Comment by dalthviz** [comment]
> Those auto-resizings are a bit odd O.o

Totally agree! Maybe those are related with the behavior I found while checking https://github.com/napari/napari/issues/5977#issuecomment-1658789366 However, adding some logic to readjust the dialog size after the `font_size` setting changes could help too 🤔 Will check!

**Comment by dalthviz** [comment]
Note: Rechecking this I ended up adding some scrollbars to the preferences pages to prevent the preference dialog changing size when the font size value is updated. Not sure if this is a good enough approach but if you have any other ideas or suggestions let me know!

**Review by psobolewskiPhD — APPROVED** [review]
I like this a lot!

**Comment by Czaki** [comment]
AS you could see on recording the font size do not behave correctly on change of current theme (in my opinion should change to theme default) 
https://github.com/napari/napari/assets/3826210/1da6c300-f469-4101-ab72-80bee3cdb212

But in the situation when the user selects the system theme and the system change theme (@psobolewskiPhD - you have this use case on mac) the font size should be preserved.

**Comment by psobolewskiPhD** [comment]
Good catch! I agree that the font size set here should apply to all themes and certainly shouldn't change when the system theme changes between light and dark.

**Comment by dalthviz** [comment]
Thanks for the feedback @psobolewskiPhD @Czaki ! So if I'm understanding correctly, things should be working something like:

pref_font_size

?

Let me know!

**Comment by Czaki** [comment]
> Thanks for the feedback @psobolewskiPhD @Czaki ! So if I'm understanding correctly, things should be working something like:

Not for me. As it will always hide the font size provided by the theme. So it may lead to strange-looking themes when someone tests. I prefer to reset the size to default for a given theme when changing theme. 

Because of this, I prefer to not mutate the `Theme` object, but apply these changes somehow in  `get_stylesheet` nhttps://github.com/napari/napari/blob/787fc5f450aeb1d7ea23a7915f096fd5185f40b2/napari/_qt/qt_resources/__init__.py#L14

**Comment by psobolewskiPhD** [comment]
I see Czaki's point. Maybe we can override the theme, but maybe have a way to return to the theme default? I guess the reset everything to default would do that.

**Comment by dalthviz** [comment]
Thanks both for the feedback! So, not mutating the theme makes sense to me :+1: I think something like adding a new `kwarg` to the `get_stylesheet` function could help with that:

```python
def get_stylesheet(
    theme_id: Optional[str] = None,
    override_variables: Optional[Dict[str, str]] = None,
    extra: Optional[List[str]] = None,
) -> str:
    """Combine all qss files into single, possibly pre-themed, style string.

    Parameters
    ----------
    theme : str, optional
        Theme to apply to the stylesheet. If no theme is provided, the returned
        stylesheet will still have ``{{ template_variables }}`` that need to be
        replaced using the :func:`napari.utils.theme.template` function prior
        to using the stylesheet.
    override_variables : dict, optional
        Dictionary of variables values that replace default theme values
    extra : list of str, optional
        Additional paths to QSS files to include in stylesheet, by default None

    Returns
    -------
    css : str
        The combined stylesheet.
    """
    stylesheet = ''
    for key in sorted(STYLES.keys()):
        file = STYLES[key]
        with open(file) as f:
            stylesheet += f.read()
    if extra:
        for file in extra:
            with open(file) as f:
                stylesheet += f.read()

    if theme_id:
        from napari.utils.theme import get_theme, template

        theme_dict = get_theme(theme_id).to_rgb_dict()
        if override_variables:
            theme_dict.update(override_variables)

        return template(stylesheet, **theme_dict)

    return stylesheet
```

However, I'm not sure how to make the preferences widget to now follow the default theme value font size when changing themes 🤔 Seems like the connection between preferences widgets and the settings is done only in one direction, right? So when the generated widget for a setting changes its value, the settings are updated but not the other way around?:

https://github.com/napari/napari/blob/787fc5f450aeb1d7ea23a7915f096fd5185f40b2/napari/_qt/dialogs/preferences_dialog.py#L122-L128

**Comment by Czaki** [comment]
It is to complex to explain all the changes that I found required, so I pushed commit with these changes. 

In general, I implement opposite connections from settings to widgets. It is not pretty, but otherwise, it will require more profound changes in the settings dialog. If you think it is worth, I may do it in separate PR.

**Comment by dalthviz** [comment]
> It is to complex to explain all the changes that I found required, so I pushed commit with these changes.
>
> In general, I implement opposite connections from settings to widgets. It is not pretty, but otherwise, it will require more profound changes in the settings dialog. If you think it is worth, I may do it in separate PR.

Thanks for the help with that! Checking locally I can see the reset behavior working :+1:

pref_font_size_reset

Regarding the changes, seems okay to me to include them here if the desire behavior for the preference is the one above, However, should be added some validation inside the `update_widget_state` function?. Checking locally, if I do something like:

```python
def update_widget_state(name, widget):
    def _update_widget_state(event):
        if name == "font_size":
            widget.state = {name: event.value}

    return _update_widget_state
``` 

I don't see errors when closing napari. Without that I got a `KeyError: 'window_position'` and seems like in the tests something similar is happening but with the `preferences_size` setting: `E           KeyError: 'preferences_size'`

Maybe the `update_widget_state` should be used only for the `font_size` setting for the moment or some settings names should be excluded?

**Comment by Czaki** [comment]
> I don't see errors when closing napari. Without that I got a `KeyError: 'window_position'` and seems like in the tests something similar is happening but with the `preferences_size` setting: `E KeyError: 'preferences_size'`
> 
> Maybe the `update_widget_state` should be used only for the `font_size` setting for the moment or some settings names should be excluded?

The source of the problem is that connection happens also to elements excluded from settings. I have pushed changes that fix this.

**Comment by dalthviz** [comment]
Note: Accidentally force pushed the branch 😓 Trying to revert it!

**Comment by Czaki** [comment]
When I go back to home i will fix.

**Comment by dalthviz** [comment]
> When I go back to home i will fix.

I think I managed to readd the commit but any extra check is appreciated! Sorry for the noise there and thanks!

**Comment by psobolewskiPhD** [comment]
It's not clear from the video, but the current state of this PR makes it that when you change `system` it will affect both the light and dark versions in sync with your OS—but not live, if your OS changes while napari is running.
If you change the current viewer theme manually, e.g. via keybind, and then font size will match the new theme, but relaunch will go back the `system` setting.

I think I'm OK with this behavior, so this is just a comment and not a blocker: 
In the long run it may be more intuitive to have the font size setting be an override. I doubt someone wants the text size to be couple to color choice. I guess maybe font size shouldn't be part of the theme after all but it's own preference?

**Comment by dalthviz** [comment]
> I guess maybe font size shouldn't be part of the theme after all but it's own preference?

I have been thinking about this too 🤔 Maybe the theme could still have a font size value but used only if the font size setting has not been changed by the user? So some possible scenarios that I have been thinking:

* Using a theme without changing the font size setting and selecting/triggering the use of a different theme: In this case makes sense to me to use the new selected theme (default) font size value.

* Changing the font size setting and now setting a different theme: In this case, since you have a different font size value from the initial one the theme provides, I think you probably like to use the interface with a specific font size value, so that should not be changed.
 
Something else I have been thinking of is to maybe add another setting to change the usage of the theme font size value. Some sort of checkbox in the same preferences page, something like `Reset font size on theme change` (although this could be kind of like adding more complexity to the user interaction with the settings?), to cover possible usages of the theme font size value.

Following the scenarios above, I think the theme font size value, if preserved, could be better used more like a fallback than a value to always follow? What do you think @psobolewskiPhD @Czaki ?

> If you change the current viewer theme manually, e.g. via keybind, and then font size will match the new theme, but relaunch will go back the system setting.

Not sure if related to this but I noticed that toggling the theme with the keybinding doesn't affect the theme setting value. Is that the expected behavior or that is something that should be fixed? Running from `main` on Windows I can see the following:

<details>

theme_toggle

</details>

**Comment by Czaki** [comment]
Not update settings when toggle via keybind is inentional.

Default font size needs to live in theme. It is critical to think about design.

I don't like checkbox idea. Maybe "reset font size" button?

**Comment by dalthviz** [comment]
Added a commit exploring the button alternative and a conditional font size change following the selected theme font size value only if the font size setting has the same value as the previous theme default font size value. A preview of the behavior checking with an example custom `blue` theme that sets its default theme value to `10pt`:

theme_toggle

Still not sure about what the behavior should be when toggling the theme via the shortcut 🤔 but could make sense to just not mess with the settings at all an then just apply the theme change with whatever value the font size setting is set (which could be the default font size value of a different theme)? So following the example with the `blue` theme you would see something like:

theme_toggle_shortcut

What do you think?

**Comment by Czaki** [comment]
Please update `QListView` stylesheet (in `00_base.qss`) to add font size to it to properly update when change font size.

**Comment by dalthviz** [comment]
> Please update QListView stylesheet (in 00_base.qss) to add font size to it to properly update when change font size.

Done! And also, thinking about the layer list, the code at the `QtLayerList` definition related to font/font-size should be removed too?:

https://github.com/napari/napari/blob/2748e226f4e46539fcfcab0bf552ce26f7848543/napari/_qt/containers/qt_layer_list.py#L59-L61

**Comment by Czaki** [comment]
> Done! And also, thinking about the layer list, the code at the `QtLayerList` definition related to font/font-size should be removed too?:

yes

**Review by psobolewskiPhD — APPROVED** [review]
This is really sweet now. Thanks @dalthviz

File: napari/settings/_appearance.py [hunk 1]
```diff
@@ -13,6 +16,13 @@ class AppearanceSettings(EventedModel):
         description=trans._("Select the user interface theme."),
         env="napari_theme",
     )
+    font_size: int = Field(
+        12 if sys.platform == 'darwin' else 9,
+        title=trans._("Font size"),
```
Comment (Czaki) [inline napari/settings/_appearance.py]: Could we read this from the `Dark` theme?

**Review by Czaki — COMMENTED** [review]
single small change

**Review by Czaki — APPROVED** [review]

**Comment by Carreau** [comment]
This got the "Ready-to-merge" label 3 days ago and got no objections. The change since label are small enough, and last approval was yesterday, so merging to move the queue forward. 

Thanks.

**Pull request merged**